src/test/java/com/mongodb/hibernate/jdbc/MongoResultSetTests.java

```diff
@@ -0,0 +1,170 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.hibernate.jdbc;
+
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import com.mongodb.client.MongoCursor;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.Calendar;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Stream;
+import org.bson.BsonDocument;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+class MongoResultSetTests {
+    private static final List<String> FIELDS = List.of("id", "title", "author");
+
+    @Mock
+    private MongoCursor<BsonDocument> mongoCursor;
+
+    private MongoResultSet mongoResultSet;
+
+    @BeforeEach
+    void setUp() {
+        mongoResultSet = new MongoResultSet(mongoCursor, FIELDS);
+    }
+
+    @Nested
+    class CloseTests {
+
+        @BeforeEach
+        void setUp() throws SQLException {
+            mongoResultSet.close();
+        }
+
+        @Test
+        @DisplayName("No-op when 'close()' is called on a closed MongoResultSet")
+        void testNoopWhenCloseStatementClosed() throws SQLException {
+            // when && then
+            assertDoesNotThrow(() -> mongoResultSet.close());
+        }
+
+        @ParameterizedTest(name = "SQLException is thrown when \"{0}\" is called on a closed MongoResultSet")
+        @MethodSource("getMongoResultSetMethodInvocationsImpactedByClosing")
+        void testCheckClosed(String label, ResultSetMethodInvocation methodInvocation) throws SQLException {
+            // when && then
+            var exception = assertThrows(SQLException.class, () -> methodInvocation.runOn(mongoResultSet));
+            assertEquals("MongoResultSet has been closed", exception.getMessage());
+        }
+
+        private static Stream<Arguments> getMongoResultSetMethodInvocationsImpactedByClosing() {
+            return Map.<String, ResultSetMethodInvocation>ofEntries(
+                            Map.entry("next()", ResultSet::next),
+                            Map.entry("wasNull()", ResultSet::wasNull),
+                            Map.entry("getString(int)", rs -> rs.getString(1)),
+                            Map.entry("getBoolean(int)", rs -> rs.getBoolean(1)),
```

Comment:
How do we handle cases where the field is not the expected type? Do we intend to have any check for this?

Reply:
I expect an exception is best, I am not sure what should be done, or what the implications are. It seems like an open design question. What have other schemaless implementations done?

Reply:
I covered this important cases (if type mismatches SQLException should be thrown, not RuntimeException) by the newly added testing case.

Reply:
Let's document this in a list of "closed questions" in the design doc? (that is, throw exceptions when wrong type).

Reply:
I think it is kind of "open question". Yeah, we have little choice but throwing `SQLException` when store bson data type is String but numeric getter method is invoked.

But how about the following scenarios:

- storage bson type is `BsonDouble` with value above float range, then a getFloat() method is called?
- storage bson type is `BsonLong`, then a getInt() method is called?

the first question is related to the case BsonType is the same (we do use `BsonDouble` to store float value); while the second is related to the case BsonType is different.

Those are all hard questions without easy answer.

Reply:
This has been documented in the design doc.